exonum/src/explorer/mod.rs

```diff
@@ -0,0 +1,976 @@
+        HeightRange(Bound::Included(range.start), Bound::Excluded(range.end))
+    }
+}
+
+impl From<RangeFrom<Height>> for HeightRange {
+    fn from(range: RangeFrom<Height>) -> HeightRange {
+        HeightRange(Bound::Included(range.start), Bound::Unbounded)
+    }
+}
+
+impl From<RangeTo<Height>> for HeightRange {
+    fn from(range: RangeTo<Height>) -> HeightRange {
+        HeightRange(Bound::Unbounded, Bound::Excluded(range.end))
+    }
+}
+
+impl HeightRange {
+    /// Ending height of the range (exclusive), given the a priori max height.
+    fn end_height(&self, max: Height) -> Height {
+        use std::cmp::min;
+
+        let inner_end = match self.1 {
+            Bound::Included(height) => height.next(),
+            Bound::Excluded(height) => height,
+            Bound::Unbounded => max.next(),
+        };
+
+        min(inner_end, max.next())
+    }
+
+    fn start_height(&self) -> Height {
+        match self.0 {
+            Bound::Included(height) => height,
+            Bound::Excluded(height) => height.next(),
+            Bound::Unbounded => Height(0),
+        }
+    }
+}
+
+/// Information about a block in the blockchain.
+///
+/// # Examples
+///
+/// ```
+/// # #[cfg(feature = "doctests")] fn main() {
+/// # use exonum::explorer::{BlockchainExplorer, BlockInfo};
+/// # use exonum::explorer::tests::sample_blockchain;
+/// # use exonum::helpers::Height;
+/// let blockchain = // ...
+/// #                sample_blockchain();
+/// let explorer = BlockchainExplorer::new(&blockchain);
+/// let block: BlockInfo = explorer.block(Height(1)).unwrap();
+/// assert_eq!(block.height(), Height(1));
+/// assert_eq!(block.len(), 3);
+///
+/// // Iterate over transactions in the block
+/// for tx in &block {
+///     println!("{:?}: {:?}", tx.location(), tx.content());
+/// }
+/// # } // main
+/// # #[cfg(not(feature = "doctests"))] fn main() {}
+/// ```
+///
+/// # JSON presentation
+///
+/// ```
+/// # #[macro_use] extern crate serde_json;
+/// # extern crate exonum;
+/// # #[cfg(feature = "doctests")] fn main() {
+/// # use exonum::helpers::Height;
+/// # use exonum::explorer::{BlockchainExplorer, BlockInfo};
+/// # use exonum::explorer::tests::sample_blockchain;
+/// # let blockchain = sample_blockchain();
+/// # let explorer = BlockchainExplorer::new(&blockchain);
+/// let block: BlockInfo = // ...
+/// #                      explorer.block(Height(1)).unwrap();
+/// assert_eq!(
+///     serde_json::to_value(&block).unwrap(),
+///     json!({
+///         // `Block` representation
+///         "block": block.block(),
+///         // Array of `Precommit`s
+///         "precommits": *block.precommits(),
+///         // Array of transaction hashes
+///         "txs": *block.transaction_hashes(),
+///     })
+/// );
+/// # } // main
+/// # #[cfg(not(feature = "doctests"))] fn main() {}
+/// ```
+#[derive(Debug)]
+pub struct BlockInfo<'a> {
+    explorer: &'a BlockchainExplorer<'a>,
+    block: Block,
```

Comment:
Imo, it would make sense to rename this field header, but it would break backward compatibility with JSON serialization. A possible compromise is renaming the field in the serialization code.